DEP: Expire deprecation of dtype/signature allowing instances

seberg opened the pull request:
We never really allowed instances here and deprecated it since NumPy 1.21 (it just failed completely in 1.21.0).

<!--         ----------------------------------------------------------------
                MAKE SURE YOUR PR GETS THE ATTENTION IT DESERVES!
                ----------------------------------------------------------------

*  FORMAT IT RIGHT:
      https://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message

*  IF IT'S A NEW FEATURE OR API CHANGE, TEST THE WATERS:
      https://www.numpy.org/devdocs/dev/development_workflow.html#get-the-mailing-list-s-opinion

*  HIT ALL THE GUIDELINES:
      https://numpy.org/devdocs/dev/index.html#guidelines

*  WHAT TO DO IF WE HAVEN'T GOTTEN BACK TO YOU:
      https://www.numpy.org/devdocs/dev/development_workflow.html#getting-your-pr-reviewed
-->

seberg committed:
DEP: Expire deprecation of dtype/signature allowing instances

We never really allowed instances here and deprecated it since
NumPy 1.21 (it just failed completely in 1.21.0).

seberg committed:
DOC: Add release note for expired dtype=/signature= ufunc deprecation

charris commented on doc/release/upcoming_changes/22540.expired.rst:
```diff
@@ -0,0 +1,3 @@
+* Passing dtype instances other than the default ones to
+  ``dtype=`` or ``signature=` in ufuncs will now raise a
```
"default" could use some clarification. I assume some of the outdate alternate (example?) types are deprecated?

seberg replied:
Hmmm, tried to make it more specific.  Basically, I still allow the singleton instances (and things that are equivalent to it).  But really, we only enforce the DType (class/type) anyway, and are just generous about allowing those.

rossbar replied:
FWIW the new wording LGTM and matches the terminology (i.e. "canonical") in the relevant NEPs

seberg committed:
DOC: Try to clarify which dtype instaces are still allowed

We still allow the "singleton" instances (which mainly applies to
our own dtypes), mainly because it wouldn't really do much good to
disallow them, even if they are not specific (since we don't enforce
the byte-order, but we never return non-native byte order for example).

rossbar commented on doc/release/upcoming_changes/22540.expired.rst:
```diff
@@ -0,0 +1,5 @@
+* Passing dtype instances other than the canonical (mainly native byte-order)
+  ones to ``dtype=`` or ``signature=` in ufuncs will now raise a ``TypeError``.
```
```suggestion
  ones to ``dtype=`` or ``signature=`` in ufuncs will now raise a ``TypeError``.
```

rossbar reviewed (approved):
LGTM modulo an rst formatting nit in the release note. Thanks @seberg !

charris committed:
MAINT: Update doc/release/upcoming_changes/22540.expired.rst

Co-authored-by: Ross Barnowski <rossbar@berkeley.edu>

charris commented:
Thanks Sebastian,
